RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,177 @@
+
+The process followed during a function/method call would be equivalent to:
+
+1. The dispatcher is called to extract the `Tensor`-likes.
+2. All `Tensor`-likes are checked for `__torch_function__`. If none exist, the internal implementation is called, and the final result is returned.
+3. A collection of types that implement `__torch_function__` is created, with no guaranteed order other than that subclasses come before superclasses.
+4. For one argument of each type, `__torch_function__` is called. The first such function or method to return something other than `NotImplemented` will be the final result. All exceptions will be propagated upward.
+5. If all `__torch_function__` implementations return `NotImplemented`, a `TypeError` is raised with an appropriate error message.
+
+In practice, for most PyTorch functions, the list of tensor-likes is already available and the dispatcher doesn't need to be called. Additionally, while equivalent to the code above, if the `Tensor`-likes are all `Tensor`s or classes for which `__torch_function__ is torch.Tensor.__torch_function__`, the internal implementation is called immediately. This is done as a performance optimisation to avoid overhead for concrete `Tensor` objects.
+
+To access super, one would do the following:
+```python
+class SubTensor(torch.Tensor):
+    def __torch_tensor__(self, func, types, args, kwargs):
+        # Pre-processing here
+        val = super().__torch_function__(
+            func,
+            tuple(
+                t for t in types if not issubclass(t, SubTensor)
+            ),
+            args,
+            kwargs
+        )
+        # Post processing here
+```
+
+This way `__torch_function__` knows the list of types to dispatch to, and it will _not_ dispatch to `SubTensor` anymore in this example.
+
+We will also recommend that all `Tensor` subclasses make their own methods go through `__torch_function__` via a decorator `@torch_function_dispatch`. This decorator was added and then removed for performance reasons, however it will be added back to allow external libraries to interface with the protocol. It will take a single argument: a dispatcher, i.e. a callable that returns an iterable of all the "duck-Tensors", or possible candidates for classes that may implement `__torch_function__`.
+
+If a library forgets to add the aforementioned decorator, then the method will no longer dispatch at all to any form of `__torch_function__`. In other words, it will lose support for the protocol.
```

Comment:
Most users won't care about losing support for the protocol. However, I think the (unstated) implication is that the returned tensor is a `Tensor` instance rather than a subclass instance. This should be stated clearly here.